modules/code-deployment/src/main/java/org/apache/ignite/internal/deployunit/StaticUnitDeployer.java

```diff
@@ -0,0 +1,176 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.deployunit;
+
+import static java.util.Collections.emptyList;
+import static org.apache.ignite.deployment.version.Version.parseVersion;
+import static org.apache.ignite.internal.deployunit.DeploymentStatus.DEPLOYED;
+import static org.apache.ignite.internal.util.CompletableFutures.allOf;
+
+import java.io.IOException;
+import java.nio.file.FileVisitResult;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.BiConsumer;
+import org.apache.ignite.deployment.version.Version;
+import org.apache.ignite.internal.deployunit.exception.DeploymentUnitReadException;
+import org.apache.ignite.internal.deployunit.metastore.DeploymentUnitStore;
+import org.apache.ignite.internal.deployunit.metastore.status.UnitNodeStatus;
+import org.apache.ignite.internal.logger.IgniteLogger;
+import org.apache.ignite.internal.logger.Loggers;
+
+/**
+ * Observes a predefined directory with statically provisioned deployment units and
+ * registers their presence in the deployment store.
+ *
+ * <p>The observer scans the {@code deploymentUnitsRoot} directory, expecting the following structure:
+ * {@code <unitId>/<version>/}
+ * For every discovered {@code unitId}-{@code version} pair that is not yet registered for this node,
+ * the observer creates (or reuses) a cluster status and then creates the node status marked as
+ * {@link DeploymentStatus#DEPLOYED}.
+ */
+public class StaticUnitDeployer {
+    private static final IgniteLogger LOG = Loggers.forClass(StaticUnitDeployer.class);
+
+    private final DeploymentUnitStore deploymentUnitStore;
+
+    private final String nodeName;
+
+    private final Path deploymentUnitsRoot;
+
+    /**
+     * Constructor.
+     */
+    public StaticUnitDeployer(
+            DeploymentUnitStore deploymentUnitStore,
+            String nodeName,
+            Path deploymentUnitsRoot
+    ) {
+        this.deploymentUnitStore = deploymentUnitStore;
+        this.nodeName = nodeName;
+        this.deploymentUnitsRoot = deploymentUnitsRoot;
+    }
+
+    /**
+     * Scans the filesystem for statically deployed units and registers their cluster and node statuses
+     * if they are not yet present in the store.
+     *
+     * <p>Already registered unit versions for this node are skipped. New ones are registered as DEPLOYED.
+     */
+    public CompletableFuture<Void> searchAndDeployStaticUnits() {
+        LOG.info("Start search static deployment units.");
```

Comment:
```suggestion
        LOG.info("Static deployment unit lookup started.");
```

Or maybe remove this log message?